tox_core/src/net_crypto/mod.rs

```diff
@@ -1021,21 +1047,27 @@ impl NetCrypto {
 
     /// Send `CookieRequest` or `CryptoHandshake` packet if needed depending on
     /// connection status and update sent counter
-    fn send_status_packet(&self, connection: &mut CryptoConnection) -> impl Future<Output = Result<(), SendPacketError>> + Send {
-        match connection.packet_to_send() {
+    fn send_status_packet(&self, connection: Arc<RwLock<CryptoConnection>>, status: Option<ConnectionStatus>)
+        -> impl Future<Output = Result<(), SendPacketError>> + Send {
+        if let Some(status) = status {
+            connection.write().status = status;
+        }
+        let packet_to_send = connection.write().packet_to_send();
+        match packet_to_send {
             Some(packet) => Either::Left(self.send_packet(packet.into(), connection)),
             None => Either::Right(future::ok(())),
         }
     }
 
     /// Send `CryptoData` packet if the connection is established.
-    fn send_data_packet(&self, connection: &mut CryptoConnection, data: Vec<u8>, packet_number: u32)
+    fn send_data_packet(&self, connection: Arc<RwLock<CryptoConnection>>, data: Vec<u8>, packet_number: u32)
         -> impl Future<Output = Result<(), SendDataError>> + Send {
-        let packet = match connection.status {
+        let buffer_start = connection.read().recv_array.buffer_start;
+        let packet = match connection.write().status {
```

Comment:
I don't really like how it looks. First we acquire read lock, then release it and acquire write lock... Need to check whether we can deadlock this way.

Reply:
Firs I got totally deadlock but then I inspect an all deals with lock acqure line-by-line and got succes. But yes, it's not seems good for me anyway: too much places wants to modify Connection state etc